### PR TITLE
fix: custom boolean types in conditional statements

### PR DESCRIPTION
**Pull request by nettijoe96**
Resolves https://github.com/gnolang/gno/issues/1087

Preprocess conditional expr with `checkOrConvertBoolType`

Unit test for `if` and `for` condition and `integ` test for invalid assignment. The invalid assignment is a case that [fails](https://go.dev/play/p/P631q105QW3) to compile in Go, but succeeded in Gno before this change. 

Run tests:
```
go test -run ^TestRunApp$ github.com/gnolang/gno/gnovm/cmd/gno 
go test -run ^TestFiles$ github.com/gnolang/gno/gnovm/tests
```

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [X] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Comment by thehowl** [comment]
Hey, sorry for the wait. I'll take a look today!

File: gnovm/pkg/gnolang/preprocess.go [hunk 1]
```diff
@@ -1701,13 +1701,23 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
...
 			case *IfStmt:
-				// Cond consts become bool *ConstExprs.
-				checkOrConvertType(store, last, &n.Cond, BoolType, false)
+				if nx, ok := (n.Cond).(*NameExpr); ok {
+					// Cond is named type, check if bool
+					checkNameExprType(store, last, nx, BoolType)
+				} else {
+					// Cond consts become bool *ConstExprs.
+					checkOrConvertType(store, last, &n.Cond, BoolType, false)
+				}
```
Comment (thehowl) [inline gnovm/pkg/gnolang/preprocess.go:1720]: I tend to think, actually, that if we change the original checkOrConvertType into [`checkOrConvertBoolType`](https://github.com/gnolang/gno/pull/2140/files#r1618434979) we can both optimise the processing and fix this issue.

Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go:1720]: I think this should be the way to go.

File: gnovm/pkg/gnolang/preprocess.go [hunk 2]
```diff
@@ -2713,6 +2723,16 @@ func checkType(xt Type, dt Type, autoNative bool) {
 		dt.String()))
 }
 
+// Checks type of NameExpr
+func checkNameExprType(store Store, last BlockNode, nx *NameExpr, t Type) {
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:2727]: The naming is not very directional, but considering the discussion above, the content of this comment can be ignored.

File: gnovm/pkg/gnolang/preprocess.go (hunk 1 above)
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go:1720]: this should be the way to go.

Comment (nettijoe96) [reply gnovm/pkg/gnolang/preprocess.go:1720]: sounds good!

File: gnovm/pkg/gnolang/preprocess.go [hunk 3]
```diff
@@ -2949,6 +2947,31 @@ func checkIntegerType(xt Type) {
...
+		xt := evalStaticTypeOf(store, last, Expr(nx))
+		if dxt, ok := xt.(*DeclaredType); ok {
+			checkType(dxt.Base, BoolType, false)
+		} else {
+			panic(fmt.Sprintf("expected declared type, but got %v", xt.Kind()))
+		}
+	} else if x != nil {
+		xt := evalStaticTypeOf(store, last, x)
+		checkBoolType(xt)
+	}
```
Comment (thehowl) [inline gnovm/pkg/gnolang/preprocess.go:2964]: actually should probaby panic in case x == nil

File: gnovm/pkg/gnolang/preprocess.go [hunk 4]
```diff
@@ -2949,6 +2947,31 @@ func checkIntegerType(xt Type) {
...
+	if cx, ok := x.(*ConstExpr); ok {
+		convertConst(store, last, cx, BoolType)
+	} else if nx, ok := (x).(*NameExpr); ok {
+		xt := evalStaticTypeOf(store, last, Expr(nx))
+		if dxt, ok := xt.(*DeclaredType); ok {
+			checkType(dxt.Base, BoolType, false)
+		} else {
+			panic(fmt.Sprintf("expected declared type, but got %v", xt.Kind()))
+		}
+	} else if x != nil {
```
Comment (thehowl) [inline gnovm/pkg/gnolang/preprocess.go:2961]: Could we not just use `t := evalStaticTypeOf(x); checkType(baseOf(t), BoolType, false)` (and avoid creating `checkBoolType`?)?

File: gnovm/pkg/gnolang/preprocess.go (hunk 3 above)
Comment (nettijoe96) [reply gnovm/pkg/gnolang/preprocess.go:2964]: Good point. Maybe `checkOrConvertIntegerType` should have this too?

**Comment by nettijoe96** [comment]
Note: currently looking into why `go test -timeout 30s -run ^TestFiles$ github.com/gnolang/gno/gnovm/tests` is failing

**Comment by thehowl** [comment]
> Note: currently looking into why `go test -timeout 30s -run ^TestFiles$ github.com/gnolang/gno/gnovm/tests` is failing

It seems like the expression can be nil, like in `for` loops where the conditional is optional. Let's remove the nil check.

**Review by thehowl — APPROVED** [review]

File: gnovm/pkg/gnolang/preprocess.go [hunk 5]
```diff
@@ -3039,6 +3037,16 @@ func checkIntegerType(xt Type) {
 	}
 }
 
+// like checkOrConvertType() but for bool type.
+func checkOrConvertBoolType(store Store, last BlockNode, x Expr) {
+	if cx, ok := x.(*ConstExpr); ok {
+		convertConst(store, last, cx, BoolType)
+	} else if x != nil {
+		t := evalStaticTypeOf(store, last, x)
+		checkType(baseOf(t), BoolType, false)
```
Comment (jaekwon) [inline gnovm/pkg/gnolang/preprocess.go:3046]: checkOrConvertIntegerType calls checkOrConvertType here, I wonder why there is asymmetry, and if we need it.

Comment (nettijoe96) [reply gnovm/pkg/gnolang/preprocess.go:3046]: I see checkOrConvertIntegerType calling checkIntegerType here https://github.com/gnolang/gno/blob/80c9ddc00bd3a0c6a401d06968ce17ab904c97ff/gnovm/pkg/gnolang/preprocess.go#L3022

Comment (nettijoe96) [reply gnovm/pkg/gnolang/preprocess.go:3046]: > also take a look at this, which I haven't reviewed; it makes changes to dependent functions. #1426

I think `checkType` would have to be changed to `assertAssignableTo` after that PR merges. I'll add a comment

**Comment by jaekwon** [comment]
thank you! modified to use new functions.

**Pull request merged**